src/array.cr

```diff
@@ -163,14 +163,26 @@ class Array(T)
   # ary == [2, 3]    # => false
   # ```
   def ==(other : Array)
-    equals?(other) { |x, y| x == y }
+    exec_recursive_pair(:==, other) do
+      return false unless equals?(other) { |x, y| x == y }
+    end
+
+    true
   end
 
   # :nodoc:
   def ==(other)
     false
   end
 
+  # See `Object#hash(hasher)`
+  def hash(hasher)
+    # TODO: it returns same hash value when an array is recursive.
+    # While `Hash` works fine, but it is inefficient.
```

Comment:
I don't follow what you would want to do different.

Reply:
For example,

```crystal
alias Rec = Array(Rec)

a = [] of Rec
b = [] of Rec
c = [] of Rec

a << b
b << a
c << c << c
```

then, `a.hash == b.hash` and `a.hash == c.hash` even though `a == b` and `a != c`.

Avoiding hash collision, it is better that `a.hash != c.hash` if possible. However it is hard to me.